algoritmy/Final Algorithm Testing.ipynb

```diff
@@ -0,0 +1,281 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "** Build model and train it **"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "w2v_model = comms.create_w2v_tf_model(batch_size, embedding_size, vocabulary_size, num_sampled)\n",
+    "train_inputs, train_labels, embeddings, nce_weights, nce_biases, loss, optimizer, init = w2v_model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sess = tf.Session()\n",
+    "sess.run(init)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Loss at iteration 0  :  1.45273\n",
+      "Loss at iteration 1000  :  1.59382\n",
+      "Loss at iteration 2000  :  1.73079\n",
+      "Loss at iteration 3000  :  1.47814\n",
+      "Loss at iteration 4000  :  1.80039\n",
+      "Loss at iteration 5000  :  1.57665\n",
+      "Loss at iteration 6000  :  1.48287\n",
+      "Loss at iteration 7000  :  1.6787\n",
+      "Loss at iteration 8000  :  1.9216\n",
+      "Loss at iteration 9000  :  1.45983\n"
+     ]
+    }
+   ],
+   "source": [
+    "## Learn the embeddings\n",
+    "for i in range(num_iter): \n",
+    "    batch, labels = generate_batch(batch_size)\n",
+    "    _, loss_eval = sess.run([optimizer, loss], {train_inputs : batch, train_labels : labels})\n",
+    "    if i % 1000 == 0: \n",
+    "        print(\"Loss at iteration\",i, \" : \",loss_eval)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 36,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "embeddings_eval, nce_weights_eval, nce_biases_eval = sess.run([embeddings, nce_weights, nce_biases]) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "** Test the model on data ** "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 115,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Take last word from each user \n",
+    "## Try to predict it using the rest \n",
+    "test_labels = [dictionary.get(wb[-1]) for wb in word_bags_test]\n",
+    "test_batches = [[dictionary.get(w) for w in wb[:-1]] for wb in word_bags_test]\n",
+    "\n",
+    "## Remove Nones\n",
+    "remove_indicators = [tl == None or all(x is None for x in tb) for tl,tb in zip(test_labels, test_batches)]\n",
+    "\n",
+    "test_labels_filter = [tl for tl, ri in zip(test_labels, remove_indicators) if not ri]\n",
+    "test_batches_filter = [tb for tb, ri in zip(test_batches, remove_indicators) if not ri]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 116,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user_embeds = [np.mean(embeddings_eval[tb, :], axis = 0) for tb in test_batches]"
```

Comment:
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-13-112c4d382fa2> in <module>()
----> 1 user_embeds = [np.mean(embeddings_eval[tb, :], axis = 0) for tb in test_batches]

<ipython-input-13-112c4d382fa2> in <listcomp>(.0)
----> 1 user_embeds = [np.mean(embeddings_eval[tb, :], axis = 0) for tb in test_batches]

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

Reply:
Stalo se ti zřejmě to, že test dataset vyšel prázdný. To se může stát, když nemáš v testovacích datech žádné produkty z trénovacích. Což při těch 14 uživatelích a 25 produktech v datech není těžké. Jinak to funguje dobře. Na ostrých datech by se nemělo stát a vzhledem k tomu, že to je spíš věc pro naše interní použití, tak bych to neošetřoval.